soc: nordic: Add LRCCONF management

adamkondraciuk opened the pull request:
Due to the possibility of simultaneous accesess to LRCCONF registers, additional management is required.

adamkondraciuk committed:
soc: nordic: Add LRCCONF management

Due to the possibility of simultaneous accesess to LRCCONF registers,
additional management is required.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

anangl commented on drivers/clock_control/clock_control_nrf2_common.h:
```diff
@@ -79,10 +79,4 @@ struct clock_lrcconf_sink {
-/**
- * @brief Request or release lrcconf main power domain
- */
-void clock_request_lrcconf_poweron_main(struct clock_lrcconf_sink *sink);
```
Please remove also the `struct clock_lrcconf_sink` definition. There is no need to keep it as it can be replaced with just `sys_snode_t`.

anangl commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -8,7 +8,7 @@
 #include <zephyr/devicetree.h>
 #include <zephyr/drivers/clock_control/nrf_clock_control.h>
-#include <hal/nrf_lrcconf.h>
+#include <soc/nordic/common/soc_lrcconf.h>
```
```suggestion
#include <soc_lrcconf.h>
```

anangl commented on soc/nordic/common/soc_lrcconf.c:
```diff
@@ -0,0 +1,63 @@
+#include <zephyr/kernel.h>
+
+static struct k_spinlock lock;
+sys_slist_t poweron_main_list;
+sys_slist_t poweron_active_list;
```
These should be `static`.

anangl commented on soc/nordic/nrf54h/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@
 
 if(CONFIG_ARM)
   zephyr_library_sources(soc.c)
+  zephyr_library_sources(../common/soc_lrcconf.c)
```
Why is it added only for ARM cores?
I think that instead of the above entry, the following should be added in `soc/nordic/common/CMakeLists.txt` (so that nRF92 could also use it):
```
zephyr_library_sources_ifdef(NRF_PLATFORM_HALTIUM soc_lrcconf.c)
```

karstenkoenig replied:
It was correct to only have it for arm cores, or rather for local domains, as global domains don't have an NRF_LRCCONF010 instance and it breaks their builds.

anangl commented on soc/nordic/nrf54h/power.c:
```diff
@@ -9,12 +9,16 @@
+#include <soc_lrcconf.h>
 #include "pm_s2ram.h"
 
+#if !IS_ENABLED(CONFIG_SOC_NRF54H20_CPURAD)
```
```suggestion
#if !defined(CONFIG_SOC_NRF54H20_CPURAD)
```
`IS_ENABLED` should only be used in C expressions. See: https://github.com/zephyrproject-rtos/zephyr/blob/7e892f8c6fd67b678ae5a6f06a73f878e9c2377a/include/zephyr/sys/util_macro.h#L117-L118

anangl commented on soc/nordic/nrf54h/soc.c:
```diff
@@ -52,13 +60,20 @@ static void power_domain_init(void)
-				      !IS_ENABLED(CONFIG_SOC_NRF54H20_CPURAD));
-
+#if !IS_ENABLED(CONFIG_SOC_NRF54H20_CPURAD)
+	sys_snode_t *node;
+#if IS_ENABLED(CONFIG_PM) || IS_ENABLED(CONFIG_POWEROFF)
+	node = pm_sys_node_id_get();
+#else
+	node = &soc_node;
+#endif
```
This looks like a leftover from some previous approach. `node` is not used by the code below.

anangl commented on soc/nordic/nrf54h/power.h:
```diff
@@ -10,11 +10,29 @@
 void nrf_poweroff(void);
 
+/**
+ * @brief Pepare to the idle state.
```
typo: "Pepare"

anangl commented on soc/nordic/nrf54h/power.h:
```diff
@@ -10,11 +10,29 @@
+/**
+ * @brief Pepare to the idle state.
+ *
+ * This function applies required PM configuration must be performed before the idle state.
```
This sentence seems to need rephrasing.

anangl commented on soc/nordic/nrf54h/power.h:
```diff
@@ -10,11 +10,29 @@
+ *
+ * @return The pointer to the node assigned to PM module.
+ */
+sys_snode_t *pm_sys_node_id_get(void);
```
This function does not seem to be needed anymore. It looks a bit like a hack actually.

anangl commented on soc/nordic/nrf54h/Kconfig:
```diff
@@ -28,6 +28,7 @@ config SOC_NRF54H20_CPUAPP
 	select NRFS_HAS_VBUS_DETECTOR_SERVICE
 	select HAS_PM
 	select HAS_POWEROFF
+	select ARM_ON_ENTER_CPU_IDLE_HOOK
```
Shouldn't this be selected only when actually needed, so `if PM || POWEROFF`?

karstenkoenig commented on soc/nordic/nrf54h/power.c:
```diff
@@ -126,20 +125,37 @@ static void do_suspend_to_ram(void)
 {
+#if !defined(CONFIG_SOC_NRF54H20_CPURAD)
+	if (state == PM_STATE_SUSPEND_TO_IDLE) {
+		soc_lrcconf_poweron_release(&pm_node, NRF_LRCCONF_POWER_DOMAIN_0);
```
This only releases the poweron request when `PM_STATE_SUSPEND_TO_IDLE` was used to call `k_cpu_idle()` via the pm subsystem, but if the pm subsystem decided against suspending the system it will call `k_cpu_idle()` [here](https://github.com/nrfconnect/sdk-zephyr/blob/main/kernel/idle.c#L85) instead and then not call the release.
This can happen if the is no `suspend-to-idle` state defined in devicetree or it has a minimum residency time that isn't met, e.g. a `k_msleep(10)` for with this pm configuration https://github.com/nrfconnect/sdk-nrf/blob/84492d6d9264745050a018d01f4f333b285864c1/tests/benchmarks/multicore/idle/boards/nrf54h20dk_nrf54h20_cpuapp_s2ram.overlay#L12C39-L12C39

This is still way better than losing power, but needs to be followed up to have all paths covered eventually

nordic-krch replied:
there is this mechanism but it is quite ugly and requires special header with a macro. It was added as a workaround for PAN where code need to be executed immediately after WFI (so function cannot be used). I have https://github.com/zephyrproject-rtos/zephyr/pull/71667 which adds function hook symmetric to on_enter hook but PR stuck.

karstenkoenig replied:
Maybe your PR would have a clean usecase now and it could be argued for again to avoid that assembly mess?

nordic-krch commented on soc/nordic/common/soc_lrcconf.c:
```diff
@@ -0,0 +1,63 @@
+static sys_slist_t poweron_main_list;
+static sys_slist_t poweron_active_list;
+
+void soc_lrcconf_poweron_request(sys_snode_t *node, nrf_lrcconf_power_domain_mask_t domain)
```
any reason why simple reference counter is not used here? I only see that the difference is that you keep list of clients (but its not used anywhere) and allow to have non-symmetric requests (second request from the same client is not counted) but it can potentially cause some issues. Reference counter would be simpler and use less resources.

karstenkoenig replied:
At least right now there is non-symmetric requests so it couldn't be dropped

karstenkoenig replied:
I think the code is symmetric now, maybe this would benefit from an onoff manager style thing for reference counting instead of manually tracking the references?

adamkondraciuk committed:
dts: arm: nordic: Add power states for nRF54H20

Add `idle` and `s2ram` power states for nRF54H20 cpuapp and cpurad.
Also the substate `idle_cache_disable` added.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

karstenkoenig commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -64,7 +64,7 @@ static const struct clock_options {
 	STRUCT_CLOCK_CONFIG(fll16m, ARRAY_SIZE(clock_options)) clk_cfg;
 	struct onoff_client hfxo_cli;
-	struct clock_lrcconf_sink lrcconf_sink;
+	sys_snode_t lrcconf_client_node;
```
nit: `fll16m_node` might be a bit less generic?

adamkondraciuk replied:
Sounds good - fixed for `hfxo` module as well

57300 commented on soc/nordic/nrf54h/power.c:
```diff
@@ -31,23 +37,43 @@ static void suspend_common(void)
 	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO, 0);
 	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, false);
 
+#if !defined(CONFIG_SOC_NRF54H20_CPURAD)
```
Should we update these to:
```suggestion
#if !defined(CONFIG_SOC_NRF54H20_CPURAD) && !defined(CONFIG_SOC_NRF54H20_ENGB_CPURAD)
```
or:
```suggestion
#if !defined(NRF_RADIOCORE)
```

adamkondraciuk replied:
Done

karstenkoenig commented:
This latest change supporting EngB needs to be applied downstream as well

adamkondraciuk commented:
> This latest change supporting EngB needs to be applied downstream as well

https://github.com/nrfconnect/sdk-nrf/pull/18284

adamkondraciuk committed:
soc: nordic: nrf54h: disable IRQ before PM config

IRQs must be disabled before starting any procedures to prepare
for low-power states.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

anangl commented on soc/nordic/nrf54h/soc.h:
```diff
@@ -36,4 +37,14 @@
+ * @return The pointer to the node assigned to the SOC module
+ *	   used for the power domain management purposes.
+ */
+sys_snode_t *soc_pd_sys_snode_get(void);
```
This looks quite weird. Why not just use `extern sys_snode_t soc_node;` in `power.c`, for example? Is this node intended to be used by any other modules/subsystems?

bjarki-andreasen reviewed (dismissed):
With the added lrcconf API being reference tracking, one node should not be shared between modules, like `soc_pd_sys_snode_get()`. Each module creates their own `sys_snode_t` and calls `soc_lrcconf_poweron_request` independently, `soc_lrcconf_poweron_request` tracks both and ensures thread safety :)

anangl commented on soc/nordic/nrf54h/soc.c:
```diff
@@ -29,6 +30,8 @@ LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
 #define HSFLL_NODE DT_NODELABEL(cpurad_hsfll)
 #endif
 
+sys_snode_t soc_node;
```
A comment explaining that this is not static because it is also accessed from `power.c` would not hurt.

bjarki-andreasen replied:
This node really should not be shared in the first place... Instead, add an power_init() function to power.c which requests the power domain, moving the node to power.c and making it static, and call `power_init()` from soc.c

bjarki-andreasen replied:
or, just move `power_domain_init()` to power.c?

anangl replied:
This is a bit complicated, because `power.c` is only included in the build for `CONFIG_PM OR CONFIG_POWEROFF`.

adamkondraciuk replied:
How about `#ifdef`ed `power_init()` call?
